### PR TITLE
Fix ipv6 address format in Host header

### PR DESCRIPTION
**Pull request by caesar-chen**
Per discussion in [#28557](https://github.com/dotnet/corefx/issues/28557). If this change looks ok, I will fix the WinHttp as well.

Fix: [#28557](https://github.com/dotnet/corefx/issues/28557) (moved to https://github.com/dotnet/runtime/issues/25661)

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 1]
```diff
@@ -471,6 +471,31 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
...
+            string host = "[::1234]:8080";
+            string ipv6Address = "http://" + host;
+            bool connectionAccepted = false;
+
+            await LoopbackServer.CreateClientAndServerAsync(async proxyUri =>
+            {
+                using (HttpClientHandler handler = CreateHttpClientHandler())
+                using (var client = new HttpClient(handler))
+                {
+                    handler.Proxy = new WebProxy(proxyUri);
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Clever

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 2]
```diff
@@ -471,6 +471,31 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
             }
         }
 
+        [Fact]
+        public async Task GetAsync_IPv6AddressInHostHeader_CorrectlyFormatted()
+        {
+            string host = "[::1234]:8080";
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Can you make the test a theory and validate with and without a port?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 1 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Actually you [taught me this](https://github.com/dotnet/corefx/pull/28150#issuecomment-373914546). : )

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 2 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Sure.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Heh, ok, at least I'm consistent :)

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 3]
```diff
@@ -269,7 +269,8 @@ private void ConsumeFromRemainingBuffer(int bytesToConsume)
             else
             {
                 Debug.Assert(_pool.UsingProxy);
-                await WriteAsciiStringAsync(uri.IdnHost).ConfigureAwait(false);
+                await WriteAsciiStringAsync(uri.HostNameType == UriHostNameType.IPv6 ?
+                    "[" + uri.IdnHost + "]" : uri.IdnHost).ConfigureAwait(false);
```
Comment (akashkc) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: How about string interpolation instead of string concatenation manually i.e. $`"[{uri.IdnHost}]" `

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: String interpolation would be more expensive here.

Comment (akashkc) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Does string interpolation has significant performance difference than string concatenation. I believe, string interpolation has better readability.

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: In this case it'll compile down to a call to string.Concat:
```C#
string.Concat("[", uri.IdnHost, "]");
```
which will incur just a single allocation for the destination string, with the data from each constituent string blitted into it.  In contrast, the string interpolation would compile down to the equivalent of:
```C#
string.Format("[{0}]", new object[] { uri.IdnHost });
```
and would thus incur at least the destination string and an object array allocation, plus the work that has to be done inside of Format, which involves a StringBuilder.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 4]
```diff
@@ -471,11 +471,12 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
             }
         }
 
-        [Fact]
-        public async Task GetAsync_IPv6AddressInHostHeader_CorrectlyFormatted()
+        [Theory]
+        [InlineData("http://", "[::1234]")]
+        [InlineData("http://", "[::1234]:8080")]
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Nit: the scheme doesn't need to be parameterized

**Review by stephentoub — APPROVED** [review]

**Comment by caesar1995** [comment]
I was trying to add test data for `https` `ws` and `wss` as well, but those scheme will make the test to hang locally, not sure why.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 5]
```diff
@@ -471,6 +471,32 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
             }
         }
 
+        [Theory]
+        [InlineData("http://", "[::1234]")]
+        [InlineData("http://", "[::1234]:8080")]
+        public async Task GetAsync_IPv6AddressInHostHeader_CorrectlyFormatted(string scheme, string host)
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: I assume this test fails without your product change?  Just checking.

Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Yes.

**Comment by caesar1995** [comment]
To provide more confidence, by stepping into the code, I can see in `Host` header, ipv6 address is appended with `[]` for the above hanging schemes. (since no test added for them)

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 4 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Will fix.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 5 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Without product change:

```
        Assert.Contains() Failure
        Not found: Host: [::1234]
        In value:  List<String> ["GET http://::1234/ HTTP/1.1", "Host: ::1234"]
```

Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Only semi-related, is the Uri we're sending to the proxy correctly formatted, or does it also need [] around the IPv6 address?

Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: On WinHttpHandler, we will append [] around the IPv6 address.
"GET http://[::1234]/ HTTP/1.1", "Proxy-Connection: Keep-Alive", "Host: [::1234]"

On SocketsHttpHandler, currently we don't:
"GET http://::1234/ HTTP/1.1", "Host: [::1234]"

Per RFC (https://tools.ietf.org/html/rfc3986#section-3.2.2), I think we should do that for the URI we are sending as well.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 6]
```diff
@@ -269,7 +269,8 @@ private void ConsumeFromRemainingBuffer(int bytesToConsume)
             else
             {
                 Debug.Assert(_pool.UsingProxy);
-                await WriteAsciiStringAsync(uri.IdnHost).ConfigureAwait(false);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: What about the if block above? Don't we need to fix that case as well?

**Comment by geoffkizer** [comment]
> I was trying to add test data for https ws and wss as well, but those scheme will make the test to hang locally, not sure why.

That's because those schemes will try to tunnel through your proxy.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Yes, we need to fix that as well.  The fix there is more involved because it needs to happen as part of the construction of the connection key -- basically we should detect that the hostname is an IPv6 address when constructing the key, and construct the hostname using [].

You won't be able to test this using a proxy as you're doing here.  But you should at least be able to test against IPv6Loopback.  There's already a test for IP address based Uris -- GetAsync_IPBasedUri_Success.  Unfortunately it's not validating the Host header.  It should be modified to validate the Host header.  And it would be nice to add https cases to that test as well.

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Thanks for pointing this out! I will make a fix and add tests to it.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 3 above)
Comment (kasper3) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Another useful note on string.Concat vs. string.Format https://github.com/dotnet/corefx/pull/8025#discussion_r60852930 (I want "bookmark favorite comment" feature in GitHub) :octocat:

Comment (akashkc) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: @stephentoub  : Thank you for explaining it with implementation side. @kasper3  : Thank you for commenting with interesting thread.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Fix the above `if` by changing `GetConnectionKey` method.

We still need to special case this `else` statement [for `Proxy` case](https://github.com/dotnet/corefx/blob/50d4e1ce620f26ec4aebe26c121a769fc8e724ab/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L180), because for this case, we pass `null` as host, thus `_pool.HostHeaderValueBytes == null`, and here is the only place to append `[]`

**Comment by caesar1995** [comment]
@dotnet-bot test Linux x64 Release Build
@dotnet-bot test OSX x64 Debug Build

**Comment by caesar1995** [comment]
@dotnet-bot test Linux x64 Release Build
@dotnet-bot test OSX x64 Debug Build

CI hang.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs [hunk 7]
```diff
@@ -140,11 +140,12 @@ private static string ParseHostNameFromHeader(string hostHeader)
         private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri, bool isProxyConnect)
         {
             Uri uri = request.RequestUri;
+            string host = uri.HostNameType == UriHostNameType.IPv6 ? "[" + uri.IdnHost + "]" : uri.IdnHost;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs]: This is adding an allocation for every request using an IPv6 address as a host name.  Is that necessary?  Since the resulting HttpConnectionKey is used to choose the pool to use, can't this be cached on the pool instead?  And if not, can we skip doing this if proxyUri == null?

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs]: > can't this be cached on the pool instead?

I think it's possible. I didn't do that in the first place, because we will need a helper logic to determine if the address we passed in is IPv6 or not (we pass in string, not uri), and I thought the helper logic could result in additional allocation. I will try to explore this option, if not able to come up with efficient way, we may need to try the second appraoch -> skip doing this if proxyUri == null.

Comment (Tratcher) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs]: IPv6 host names are unusual, saving a few extra allocations for them is not a high priority.

**Comment by caesar1995** [comment]
@stephentoub Do you think this change will result in smaller allocation?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 8]
```diff
@@ -117,12 +117,15 @@ public HttpConnectionPool(HttpConnectionPoolManager poolManager, HttpConnectionK
 
             if (_host != null)
             {
+                bool isHostTypeIPv6 = _host.Split(':').Length - 1 > 1;
```
Comment (karelz) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Should it be just `_host.Contains(':')`? Why does it check for 2+ `:`?

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: A valid IPv6 address contains at least two `:`, so I check it for two times.

Comment (karelz) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Split allocates unnecessarily - even if just per ConnectionPool, what about using `IndexOf` on the string twice?

Comment (kasper3) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: if it's performance critical, then I think more efficient way is:

```c#
var count = 0;
var isHostTypeIPv6 = false;
foreach (char c in _host)
{
  if (c == ':' && ++count > 1)
  {
    isHostTypeIPv6 = true;
    break;
  }
}
```

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Consider `host` is not a long string, `IndexOf` can make the code cleaner.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 9]
```diff
@@ -471,6 +471,69 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
...
+
+            Assert.True(connectionAccepted);
+        }
+
+        public static IEnumerable<object[]> SecureAndNonSecure_IPBasedUri_MemberData() =>
+            from address in new[] { IPAddress.Loopback, IPAddress.IPv6Loopback }
+            from useSsl in new[] { true, false }
+            select new object[] { address, useSsl };
+
+        [OuterLoop]
```
Comment (caesar1995) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Add to outerloop since the execution for this test is long, will cause Linux catastrophic failure in innerloop run.

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 10]
```diff
@@ -117,12 +117,16 @@ public HttpConnectionPool(HttpConnectionPoolManager poolManager, HttpConnectionK
 
             if (_host != null)
             {
+                int posColon = -1;
+                bool isHostTypeIPv6 = (-1 != (posColon = _host.IndexOf(':')) && -1 != (posColon = _host.IndexOf(':', posColon)));
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Isn't the second _host.IndexOf always going to succeed, because you're passing in posColon rather than posColon+1?  Seems like we need some more tests, too, to catch this kind of thing.

Is it true that every IPv6 address contains at least two colons, and that if an address contains two colons it must be an IPv6 address?  If so, with the aforementioned change, the logic seems fine.

If it's not true, a better approach might be to augment the HttpConnectionPool ctor to accept a "Uri exemplarUri", then at the ctor call site pass in the request.Uri, and here use Uri.HostNameType instead of doing this check manually.

Comment (Tratcher) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Yes, every ipv6 contains at least two colons and nothing else does.

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Ok, then with the index fix, this looks good, e.g.
```C#
bool isHostTypeIPv6 = (posColon = _host.IndexOf(':')) != -1 && _host.IndexOf(':', posColon+1) != -1;
```

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Will fix.

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

**Comment by caesar1995** [comment]
The new test I added: [`GetAsync_SecureAndNonSecureIPBasedUri_CorrectlyFormatted`](https://github.com/dotnet/corefx/pull/28578/files#diff-06a6483ff6bb478e06f199d69522bcd4R508) will consistently causing catastrophic failure on Centos, Fedora, RedHat, and Ubuntu 1804, while passing on other distros. I think it could be different CurlHandler version causing the issue. (Since Windows run is green every time).

I will disable the test against Linux, to see if this is the issue.

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 9 above)
Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Why is the text execution long?

Why does it cause catastrophic failure on Linux?

This seems like a straightforward test, what am I missing?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 11]
```diff
@@ -117,12 +117,16 @@ public HttpConnectionPool(HttpConnectionPoolManager poolManager, HttpConnectionK
 
             if (_host != null)
             {
+                int posColon = -1;
+                bool isHostTypeIPv6 = (posColon = _host.IndexOf(':')) != -1 && _host.IndexOf(':', posColon+1) != -1;
+                string hostAddress = isHostTypeIPv6 ? "[" + _host + "]" : _host;
+
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Can we just use Uri.CheckHostName here?

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Alternatively, can we just go back to checking for HostNameType == IPv6 in GetConnectionKey, but use uri.Host instead of uri.IdnHost in that case?  I think that will return the address with [], right?

I'm kinda queasy about doing our own parsing here.  Uri does all sorts of stuff for us already, let's just use it if possible instead of trying to duplicate the logic ourselves.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 9 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: The Linux catastrophic failure confused me a lot as well. Here is my observation:

1. Before adding this test, the CI is green.
2. After I added the test, it will consistently (every time ) cause catastrophic failure on Centos, Fedora, RedHat, and Ubuntu 1804, while passing on other distros. No matter I put it in innerloop/outerloop.
3. After I disable the test on CurlHandler, no catastrophic failure with `System.Net.Http`. (There is one with `System.Runtime.Extensions.Tests`, I think it's totally unrelated).

> Why is the text execution long?

That's my guess. The [execution log](https://mc.dot.net/#/user/caesar1995/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/7db9f75494d245795a018e7c76652085703488bc/workItem/System.Net.Http.Functional.Tests/wilogs) didn't show any test failure. So I doubt the execution failed is because of hang, then timeout.

> Why does it cause catastrophic failure on Linux?

The test consistently cause catastrophic failure on Centos, Fedora, RedHat, and Ubuntu 1804, while passing on other distros. I think it could be (very likely) different CurlHandler version causing the issue. (Since Windows run is green every time).

To summary: I think it's a strange issue with CurlHandler causing the catastrophic failure. To unblock the PR, I will open a new issue to tract this, and disable the test against CurlHandler.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 11 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: I will also point out that retrieving parts of the Uri (e.g. hostname) is already a bit of a perf bottleneck.  Uri.IdnHost is allocating a string.  I believe we also end up allocating later when we retrieve path/query etc.  

We should look at optimizing this better, but not for 2.1.

In the meantime let's just do something simple and correct, and we can optimize interaction with Uri in the future.

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: > can we just go back to checking for HostNameType == IPv6 in GetConnectionKey, but use uri.Host instead of uri.IdnHost in that case?

WinHttp [clearly documented](https://github.com/dotnet/corefx/issues/28557#issuecomment-377085162) that it will only accept Punycode host name. Changing to uri.Host may bring regression for SocketsHttpHandler, and I think that's risky for 2.1.

> I think that will return the address with [], right?

Yes, uri.Host will return the address with [].

> I'm kinda queasy about doing our own parsing here.

I will change to `Uri.CheckHostName`.

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: > WinHttp clearly documented that it will only accept Punycode host name. Changing to uri.Host may bring regression for SocketsHttpHandler, and I think that's risky for 2.1.

I'm not suggesting we change it in *all* cases.  Obviously that would break Punycode handling.  Just in the case where it's an IP address.

Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: I think that should work. I will push a new commit.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs [hunk 12]
```diff
@@ -140,11 +140,12 @@ private static string ParseHostNameFromHeader(string hostHeader)
         private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri, bool isProxyConnect)
         {
             Uri uri = request.RequestUri;
+            bool isIPv6Address = uri.HostNameType == UriHostNameType.IPv6;
 
             if (isProxyConnect)
             {
                 Debug.Assert(uri == proxyUri);
-                return new HttpConnectionKey(HttpConnectionKind.ProxyConnect, uri.IdnHost, uri.Port, null, proxyUri);
+                return new HttpConnectionKey(HttpConnectionKind.ProxyConnect, isIPv6Address ? uri.Host: uri.IdnHost, uri.Port, null, proxyUri);
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs]: Nit: space before colon in conditional expression, here and below

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 3 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Since we're now using uri.Host in the GetConnectionKey path, seems like it would make sense to use it here as well for consistency and to avoid the extra alloc (maybe?).

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs (hunk 12 above)
Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs]: Oops T.T, will fix.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs [hunk 13]
```diff
@@ -140,11 +140,12 @@ private static string ParseHostNameFromHeader(string hostHeader)
         private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri, bool isProxyConnect)
         {
             Uri uri = request.RequestUri;
+            bool isIPv6Address = uri.HostNameType == UriHostNameType.IPv6;
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs]: We should add a comment here, something like:

// If the hostname is an IPv6 address, uri.IdnHost will return the address without enclosing [].
// In this case, use uri.Host instead, which will correctly enclose with [].
// Note we don't need punycode encoding if it's an IP address, so using uri.Host is fine.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 3 above)
Comment (caesar1995) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I agree, will change and add comment as well.

**Review by geoffkizer — APPROVED** [review]

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 14]
```diff
@@ -471,6 +471,74 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
...
+
+            await LoopbackServer.CreateClientAndServerAsync(async url =>
+            {
+                host = $"{url.Host}:{url.Port}";
+                using (HttpClientHandler handler = CreateHttpClientHandler())
+                using (var client = new HttpClient(handler))
+                {
+                    if (useSsl)
+                    {
+                        handler.ServerCertificateCustomValidationCallback = delegate { return true; };
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: You mentioned in a separate issue that you were seeing this test hang with CurlHandler on some OSes.  Have you tried using `HttpClientHandler.DangerousAcceptAnyServerCertificateValidator` here instead of `delegate { return true; }`?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 15]
```diff
@@ -471,6 +471,74 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
...
+                        handler.ClientCertificateOptions = ClientCertificateOption.Automatic;
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Why is this needed?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 14 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: No I didn't. I will try it and enable the test to see if this can resolve the hang.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 15 above)
Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Will remove.

**Comment by caesar1995** [comment]
@stephentoub 

Changing to `HttpClientHandler.DangerousAcceptAnyServerCertificateValidator` resolved the Linux catastrophic failure. Close #28703.

@dotnet-bot test NETFX x86 Release Build

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 16]
```diff
@@ -518,7 +518,12 @@ public static IEnumerable<object[]> GetAsync_IPBasedUri_Success_MemberData()
                 {
                     if (useSsl)
                     {
-                        handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
+                        handler.ServerCertificateCustomValidationCallback =
+#if netcoreapp
+                            HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Use TestHelper.AllowAllCertificates

Comment (caesar1995) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Will change.

**Comment by caesar1995** [comment]
Unrelated failure.

@dotnet-bot test Linux x64 Release Build
@dotnet-bot test OSX x64 Debug Build
